platform/brains2/src/lib/sync.cpp

```diff
@@ -3,20 +3,41 @@
 #include "fsl_iomuxc.h"
 #include "board.h"
 #include "duo/pins.h"
+#include <Arduino.h>
+
+#define SYNC_NVIC_PRIORITY 150
+#define SYNC_PIN_IRQ GPIO1_Combined_0_15_IRQn
+#define SYNC_PIN_IRQ_HANDLER GPIO1_Combined_0_15_IRQHandler
+
 
 namespace Sync {
+    void handleSyncInterrupt(void) {
+      GPIO_PortClearInterruptFlags(SYNC_IN_PORT, 1U << SYNC_IN_PIN);
+      Sync::syncPinState = 1;
+    }
+
     void init() {
         IOMUXC_SetPinMux(SYNC_IN_PINMUX, 0U);
         gpio_pin_config_t sync_in_config = { kGPIO_DigitalInput, 0, kGPIO_IntRisingEdge};
+        GPIO_SetPinInterruptConfig(SYNC_IN_PORT, SYNC_IN_PIN, kGPIO_IntFallingEdge);
+        NVIC_SetPriority(SYNC_PIN_IRQ, SYNC_NVIC_PRIORITY);
+        NVIC_SetVector(SYNC_PIN_IRQ, (uint32_t)&Sync::handleSyncInterrupt);
+        EnableIRQ(SYNC_PIN_IRQ);
         GPIO_PinInit(SYNC_IN_PORT, SYNC_IN_PIN, &sync_in_config);
+        GPIO_PortClearInterruptFlags(SYNC_IN_PORT, 1 << SYNC_IN_PIN);
+        GPIO_PortEnableInterrupts(SYNC_IN_PORT, 1 << SYNC_IN_PIN);
 
         IOMUXC_SetPinMux(SYNC_OUT_PINMUX, 0U);
         gpio_pin_config_t sync_out_config = { kGPIO_DigitalOutput, 0};
         GPIO_PinInit(SYNC_OUT_PORT, SYNC_OUT_PIN, &sync_out_config);
     }
 
     uint32_t read() {
-        return GPIO_PinRead(SYNC_IN_PORT, SYNC_IN_PIN);
+       if(syncPinState == 1) {
+         syncPinState = 0;
+         return HIGH;
+      }
+      return LOW;
```

Comment:
I think this implementation relies on us calling `read` often enough. A better implementation might be to increase a counter in the interrupt handler,  and have this function return how many ticks happened since last read, which should almost always be 0 or 1.

Reply:
yeah, okay that makes sense I guess. Not calling sync frequently enough shouldn't really be an issue in practice.
My thought was that the count should indicate how many sync pulses happened since last time sync was called, and that many ticks should be added to the tempo counter.
But, I think this implementation of "did the sync go high since last time" is good enough, or probably better because of the debouncing behaviour you mentioned.